### PR TITLE
BOOKKEEPER-1018: Allow client to select older V2 protocol (no protobuf)

### PR DESCRIPTION
**Pull request by govind-menon**
Tested manually - running all tests locally. Will tag contributors. when ready for review (for now putting up to test via Screwdriver and manual review)

### CODE REVIEW

**Comment by govind-menon** [comment]
Tagging @sijie and @eolivelli  for review. Will amend test cases in a follow up.

**Comment by eolivelli** [comment]
@sijie  @govind-menon @merlimat 
IMHO we should at least add a test case for the most simple cases (addEntry, readEntries, readLAC...) + basic auth

So that in the future we will not forget about this "legacy" option, as actual is not "legacy" at all if Yahoo/Pulsar is using it in production
I wonder if there is a way to automatically run the full suite with the legacy protocol and if this is really "interesting".

Issue:
should we implement all new features on the v2 protocol as well ? 
I'm thinking for instance to STARTTLS and to Long Poll Reads which have not been merged yet

**Comment by revans2** [comment]
@eolivelli I totally agree we do need some test cases for this.  The reason that we use the older API is because of the large amount of object allocation that happens with protocol buffers.  I have heard similar complains when HDFS moved to protocol buffers, but it was not fixed there either.

**Comment by eolivelli** [comment]
@revans2  I'm really OK with supporting the legacy protocol, let's find a way to guarantee the quality of that mode too.
Maybe we could add  some system property to the set suite and configure a Jenkins Job to periodically run the whole suite with the V2 protocol
@sijie  @fpj   do you think it will be possible ?

**Comment by govind-menon** [comment]
@eolivelli I can add a V2 test suite (there may be some duplication which I'll try to refactor out).

**Comment by eolivelli** [comment]
@govind-menon +1 for me maybe just cover the basic feature (addEntry, readEntries, readLAC...) + basic auth

we will need to add new testcases for future protocol changes as well mabye we should discuss this in a separate email thread

**Comment by merlimat** [comment]
@eolivelli 
> So that in the future we will not forget about this "legacy" option, as actual is not "legacy" at all if Yahoo/Pulsar is using it in production

I think that for long term we should converge to one single wire protocol that is both extensible and resource efficient. 

BK v2 protocol has its own issues as well (which is why protobuf was introduced): 
 * Fixed binary protocol, cannot add new fields and have easy upgrade
 * Doesn't have a request id. For example if a client tries to read the same entry twice, it cannot distinguish the response for each request, thus needing additional logic in the client to keep track of it.

**Comment by sijie** [comment]
@merlimat @eolivelli - what is the deal for this pr? shall we add the tests now or doing it in a separated jira?

**Comment by govind-menon** [comment]
I was thinking of doing it in a separate JIRA. @merlimat  Had suggested a way of running the existing test suite with the V2 protocol set.

**Comment by eolivelli** [comment]
+1 for separate JIRA

Overall LGTM

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 1]
```diff
@@ -803,6 +856,7 @@ void errorOutReadKey(final CompletionKey key) {
     }
 
     void errorOutReadKey(final CompletionKey key, final int rc) {
+        LOG.info("Removing completion key: {}", key);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: @govind-menon can you remove or change to  "debug" level ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 2]
```diff
@@ -835,6 +889,7 @@ void errorOutWriteLacKey(final CompletionKey key) {
     }
 
     void errorOutWriteLacKey(final CompletionKey key, final int rc) {
+        LOG.info("Removing completion key: {}", key);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: @govind-menon can you remove or change to  "debug" level ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 3]
```diff
@@ -859,6 +914,7 @@ void errorOutReadLacKey(final CompletionKey key) {
     }
 
     void errorOutReadLacKey(final CompletionKey key, final int rc) {
+        LOG.info("Removing completion key: {}", key);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: @govind-menon can you remove or change to  "debug" level ?

**Comment by eolivelli** [comment]
I'm sorry I have missed some LOG.info lines
Maybe can you @govind-menon  remove them or change to 'debug' level ?

**Comment by govind-menon** [comment]
@eolivelli @sijie Comments addressed.

**Comment by govind-menon** [comment]
@sijie Good to merge? As we discussed in the meeting the test cases and other parts of the API will be a separate PR

**Pull request closed without merging**